### ISSUE TITLE
LDAP Filter is Ignored

### ISSUE THREAD

**Issue by rlazimi-dev** [ClickHouse/ClickHouse#25686]
**Describe the unexpected behaviour**
LDAP authentication process ignores query that narrows down users in a given dn.

**How to reproduce**
in client:
```
CREATE ROLE IF NOT EXISTS admin;
GRANT ALL ON *.* to admin WITH GRANT OPTION;

CREATE ROLE IF NOT EXISTS analyst;
REVOKE ALL ON *.* FROM ALL EXCEPT admin,default;
GRANT SELECT ON *.* to analyst;
```

in server:
```
<yandex>
    <user_directories>

        <ldap>
            <server>ldap_example</server>
            <roles>
                <admin />
            </roles>
            <role_mapping>

                <base_dn>dc=example,dc=net</base_dn>
                <scope>subtree</scope>
                <search_filter>(&amp;(uid=bind_ie)(memberOf=cn=example-ie,cn=groups,cn=accounts,dc=example,dc=net))</search_filter>

                <!-- attribute is irrelevant and is not being used -->
                <attribute>dn</attribute>

            </role_mapping>
        </ldap>

    </user_directories>
</yandex>
```

ldap filter, base dn, and resulting records using command line client (shortened):

base dn: dc=example,dc=net
filter: (memberOf=cn=example-ie,cn=groups,cn=accounts,dc=example,dc=net)

```
# extended LDIF
#
# LDAPv3
# base <dc=example,dc=net> with scope subtree
# filter: (memberOf=cn=example-ie,cn=groups,cn=accounts,dc=example,dc=net)
# requesting: ALL
#

# rlazimi, users, accounts, example.net
dn: uid=rlazimi,cn=users,cn=accounts,dc=example,dc=net
memberOf: cn=example-ie,cn=groups,cn=accounts,dc=example,dc=net
uid: rlazimi


# apeavey, users, accounts, example.net
dn: uid=apeavey,cn=users,cn=accounts,dc=example,dc=net
memberOf: cn=example-ie,cn=groups,cn=accounts,dc=example,dc=net
uid: apeavey
```

My organization has one base dn –`dc=example,dc=net`- that contains all of their users, and I am trying to authenticate into clickhouse with only a specific subset of those users, which I specify using a query. I want to authenticate ldap records where `memberOf=cn=example-ie,cn=groups,cn=accounts,dc=example,dc=net`.

When I use this filter in clickhouse's ldap client, it appears as if there is no filtering occuring, because **everyone** in the base dn is authenticated and granted the `admin` role. 

**Expected behavior**
In my example, rlazimi and apeavey are the only users who will have the admin role. Anyone beside rlazimi or apeavey in my example would not be authenticated.

**Comment by traceon** [comment]
```xml
            <roles>
                <admin />
            </roles>
```
The above means: "grant all users learned form this LDAP directory `admin` role unconditionally."

---

```xml
                <!-- attribute is irrelevant and is not being used -->
                <attribute>dn</attribute>
```
Why do you think the attribute is irrelevant? During role mapping, the role name is taken from the value of this attribute.

---

```xml
                <search_filter>(&amp;(uid=bind_ie)(memberOf=cn=example-ie,cn=groups,cn=accounts,dc=example,dc=net))</search_filter>
```
`search_filter` is a filter that is used to list all groups that each user belongs to. The filter you specified 1) doesn't list groups, 2) filters only 1 hardcoded user `bind_ie`.

---

> I want to authenticate ldap records where memberOf=cn=example-ie,cn=groups,cn=accounts,dc=example,dc=net.

This is not what `search_filter` is used for.

**Comment by rlazimi-dev** [comment]
>             <roles>
>                <admin />
>            </roles>
> The above means: "grant all users learned form this LDAP directory admin role unconditionally."

How can I conditionally assign roles?

___
> This is not what search_filter is used for.

I do not have much experience with LDAP, so I assumed that the search filter narrows down the users that this role applies to.

Is it possible to authenticate only a subset of users in a given dn? if it is, then how can I do so?

___

> Why do you think the attribute is irrelevant? During role mapping, the role name is taken from the value of this attribute.

I don't think it's relevant because I already specified the `admin` role in the `<roles>` attribute. Why does there also need to be a role defined in LDAP?

**Comment by traceon** [comment]
> How can I conditionally assign roles?

The only way for doing that is running separate search for each user, and extracting role names from the results of that search. This is what `role_mapping` is used for.

---

> Is it possible to authenticate only a subset of users in a given dn? if it is, then how can I do so?

No, this is not possible. Everyone under the [constructed] `bind_dn` will be authenticated.

---

> I assumed that the search filter narrows down the users that this role applies to - as this is how the filter works in the ldap command line client.

> Why does there also need to be a role defined in LDAP?

Take a look at the docs for better understanding of the functionality: https://clickhouse.tech/docs/en/operations/external-authenticators/ldap/

**Comment by rlazimi-dev** [comment]
Let me clarify what I'm trying to achieve, as I'm not fully understanding.

**My setup**:
My company's ldap is set up so that we have one dn (dc=example,dc=net), that I use as `base_dn`, that has every user in our company. Each user has a `memberOf` attribute that defines which team they are a part of.

**What I am trying to achieve**:
In ClickHouse, I set `base_dn=dc=example,dc=net` as the context for a search filter. I then use a search filter, (memberOf=myteam), so that only the members of myteam are authenticated with the admin role for example. I would then set up another <ldap> section with a search filter of (!(memberOf=myteam)) so that all others who authenticate and who are not in my team will have a read-only role.

In short, I want an `<ldap>` section per team, with the constraint that my `base_dn` contains all users in my company each with an attribute that specifies their team.

how can I go about this? I'm unable to come up with a solution using the documentation alone.


___
> No, this is not possible. Everyone under the [constructed] bind_dn will be authenticated.

Also, here, did you mean everyone under `base_dn`

**Comment by traceon** [comment]
`base_dn` and `search_filter` are used only when searching for groups the user belongs to. You can't restrict someusers to not authenticate based on that.

What you can do is define a read-only role in clcikhouse, assign it to anyone authenticated through LDAP user directory, then setup a role mapping that would return the list of groups each uses is member of. And then define roles with corresponding names in clickhouse with extended rights (a role `myteam` that will have admin rights).

Alternatively, you can create dedicated groups for clickhouse (e.g., with `clichouse_` prefix) in LDAP and add users to those groups accordingly. Like `clickhouse_admin`, `clickhouse_readonly` etc., and make corresponding users members of corresponding groups.

**Comment by rlazimi-dev** [comment]
Hi @traceon, thank you for your help - I achieved what I wanted using your first suggestion.

So to summarize: a user-defined value is extracted using `<search_filter>`, `<attribute>`, and optionally `<prefix>`. The extracted value will be the same exact string as, typically a more privileged, role name in ClickHouse. The role with that name will be applied **on top of** the roles mentioned in the `<roles>` tag.


I suggest the documentation is slightly modified to make it clear how to authenticate all while only granting permissions to some. Perhaps you can include an (or even this) example.
